crates/holochain_types/src/app.rs

```diff
@@ -475,25 +520,63 @@ impl InstalledAppCommon {
         &self.role_assignments
     }
 
-    /// Add a cloned cell
-    pub fn add_clone(&mut self, role_id: &AppRoleId, cell_id: CellId) -> AppResult<()> {
-        let role = self.role_mut(role_id)?;
+    /// Get the next clone index for a role id.
+    pub fn next_clone_index(&self, role_id: &AppRoleId) -> AppResult<u32> {
+        Ok(self.role(role_id)?.next_clone_index)
+    }
+
+    /// Increment index for next clone cell for a role id.
+    pub fn increment_next_clone_index(&mut self, role_id: &AppRoleId) -> AppResult<()> {
+        self.role_mut(role_id)?.next_clone_index += 1;
+        Ok(())
+    }
+
+    /// Add a cloned cell.
+    pub fn add_clone(&mut self, clone_id: &CloneId, cell_id: &CellId) -> AppResult<()> {
+        let app_role_assignment = self.role_mut(&clone_id.as_base_role_id())?;
         assert_eq!(
             cell_id.agent_pubkey(),
-            role.agent_key(),
+            app_role_assignment.agent_key(),
             "A clone cell must use the same agent key as the role it is added to"
         );
-        if role.clones.len() as u32 >= role.clone_limit {
-            return Err(AppError::CloneLimitExceeded(role.clone_limit, role.clone()));
+        if app_role_assignment.clones.len() as u32 >= app_role_assignment.clone_limit {
+            return Err(AppError::CloneLimitExceeded(
+                app_role_assignment.clone_limit,
+                app_role_assignment.clone(),
+            ));
+        }
```

Comment:
Here's that check again for the third time :sweat_smile: So maybe the check should just live here, since this is the "deepest" function out of the 3? It also makes sense that this check should live beside the DuplicateCloneIds check below.

Wherever the check lives, it should be immune to race conditions. So, this seems like a good place for it, since the state would have had to be locked before even calling this function.